### PR TITLE
discord voice

### PR DESCRIPTION
**Pull request by tcm390**
related: https://github.com/ai16z/eliza/issues/244

This PR aims to improve and fix the Discord bot's voice functionality and enhance the code structure for better usability.

**1. Fix Bot Voice Singleton Issue**
Resolved an issue where the bot's voice functionality failed because the service instance was shared as a singleton across different services. Introduced a Map to ensure each service type has its own singleton instance. This guarantees only one instance of each subclass is created and reused.

**2. Add shouldRespond Function**
Added a shouldRespond function for the bot's voice feature to control when the bot should respond to user voice inputs.

**3. Refactor Transcription Process**
Refactored the transcription process by adding a debounce function, ensuring voice messages are processed only when silence is detected.

**4. Enable Bot to Respond to Text Messages in Voice Channels**
Updated the bot to handle text messages sent in voice channels.

**5. Move Template Functions to templates.ts**
Relocated template-related functions to a new templates.ts file for reusability in message.ts and voice.ts.

**6. Add Optional DISCORD_VOICE_CHANNEL_ID in .env**
Introduced a new constant, DISCORD_VOICE_CHANNEL_ID, in the .env file to allow users to specify a voice channel the bot should join.

**7. Implemented Audio Playback Interrupt Mechanism**
Add a sliding window buffer that monitors the audio volume while the agent is speaking. If the average volume of the user's audio exceeds the defined threshold, it indicates active speaking. When active speaking is detected, stop the agent's current audio playback to avoid overlap.

### CODE REVIEW

**Review by lalalune — APPROVED** [review]

File: packages/client-discord/src/voice.ts [hunk 1]
```diff
@@ -361,237 +394,332 @@ export class VoiceManager extends EventEmitter {
...
+    }
+
+    private async _shouldRespond(
+        message: string,
+        userId: UUID,
+        channel: BaseGuildVoiceChannel,
+        state: State
+    ): Promise<boolean> {
+        if (userId === this.client.user?.id) return false;
+        // if (message.author.bot) return false;
```
[the remark this answers is not in the document]
Comment (tcm390) [reply packages/client-discord/src/voice.ts]: updated: https://github.com/ai16z/eliza/pull/338/commits/9a154406f163c8dfbe5a78c4edf0f703bcb0ca02

File: packages/client-discord/src/messages.ts [hunk 2]
```diff
@@ -341,7 +238,7 @@ export class MessageManager {
         if (
             message.interaction ||
             message.author.id ===
-            this.client.user?.id /* || message.author?.bot*/
+                this.client.user?.id /* || message.author?.bot*/
```
[the remark this answers is not in the document]
Comment (tcm390) [reply packages/client-discord/src/messages.ts]: updated: https://github.com/ai16z/eliza/pull/338/commits/9a154406f163c8dfbe5a78c4edf0f703bcb0ca02

File: packages/client-discord/src/voice.ts [hunk 3]
```diff
@@ -361,237 +394,332 @@ export class VoiceManager extends EventEmitter {
...
+                        userName
+                    );
+                }, DEBOUNCE_TRANSCRIPTION_THRESHOLD);
+            } catch (error) {
+                console.error(
+                    `Error processing buffer for user ${userId}:`,
+                    error
+                );
+            }
+        };
```
[the remark this answers is not in the document]
Comment (tcm390) [reply packages/client-discord/src/voice.ts:435]: updated https://github.com/ai16z/eliza/pull/338/commits/169fd72cc77a87210154b6e51f3f1abeba49ec26

File: packages/client-discord/src/messages.ts [hunk 4]
```diff
@@ -88,108 +87,6 @@ export type InterestChannels = {
     };
 };
 
-const discordShouldRespondTemplate =
```
[the remark this answers is not in the document]
Comment (tcm390) [reply packages/client-discord/src/messages.ts:91]: Thank you for the review! I’m not entirely sure what you mean by "similar templates," as I didn’t see any others explicitly named as templates. However, I did notice some export functions in messages.ts and voice.ts, so I moved them to utils.ts for better organization. Could you confirm if that aligns with what you were referring to?

https://github.com/ai16z/eliza/pull/338/commits/00a64dc571c26ad4e99c8b50f2509b3a0da64164

https://github.com/ai16z/eliza/pull/338/commits/1f2a7f3f802662bf1c05c3858a246b2361eb6618

File: packages/client-discord/src/messages.ts [hunk 5]
```diff
@@ -23,304 +19,33 @@ import {
     UUID,
 } from "@ai16z/eliza/src/types.ts";
 import { stringToUuid } from "@ai16z/eliza/src/uuid.ts";
```
Comment (ponderingdemocritus) [inline packages/client-discord/src/messages.ts:21]: remove fulll imports from eveyrwhere - /src/uuid.ts

**Comment by tcm390** [comment]
Closing this PR due to too many conflicts. I’ll create a new PR with a clean base to address the issues.

**Pull request closed without merging**